pkg/hub/controllers/cluster/deregister.go

```diff
@@ -0,0 +1,330 @@
+		if err != nil {
+			return err
+		}
+		cluster.Status.RegistrationStatus = status
+		cluster.Status.IsDeregisterInProgress = isDeregisterInProgress
+		return r.Status().Update(ctx, cluster)
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) updateClusterDeregisterStatus(status hubv1alpha1.RegistrationStatus, ctx context.Context, cluster *hubv1alpha1.Cluster) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := r.Get(ctx, types.NamespacedName{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+		}, cluster)
+		if err != nil {
+			return err
+		}
+		cluster.Status.RegistrationStatus = status
+		return r.Status().Update(ctx, cluster)
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) setIsDeregisterInProgress(ctx context.Context, cluster *hubv1alpha1.Cluster, isDeregisterInProgress bool) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		err := r.Get(ctx, types.NamespacedName{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+		}, cluster)
+		if err != nil {
+			return err
+		}
+		cluster.Status.IsDeregisterInProgress = isDeregisterInProgress
+		return r.Status().Update(ctx, cluster)
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func constructServiceAccount() *corev1.ServiceAccount {
+	svcAcc := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceAccountName,
+			Namespace: ControlPlaneNamespace,
+		},
+	}
+	return svcAcc
+}
+
+func (r *Reconciler) getOperatorClusterRole(ctx context.Context) (*rbacv1.ClusterRole, error) {
+	operatorClusterRole := &rbacv1.ClusterRole{}
+	err := r.MeshClient.Get(ctx, types.NamespacedName{Name: operatorClusterRoleName}, operatorClusterRole)
+	if err != nil {
+		return nil, err
+	}
+	return operatorClusterRole, nil
+}
+
+func constructClusterRole(clusterRole *rbacv1.ClusterRole) *rbacv1.ClusterRole {
+	clusterRoleCopy := &rbacv1.ClusterRole{
+		TypeMeta:   metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{Name: clusterRoleName, Namespace: ControlPlaneNamespace},
+		Rules:      clusterRole.Rules,
+	}
+
+	return clusterRoleCopy
+}
+
+func constructClusterRoleBinding() *rbacv1.ClusterRoleBinding {
+	clusterRole := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterRoleBindingName,
+			Namespace: ControlPlaneNamespace,
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "ClusterRole",
+			Name:     clusterRoleName,
+		},
+		Subjects: []rbacv1.Subject{{
+			Kind:      "ServiceAccount",
+			Name:      serviceAccountName,
+			Namespace: ControlPlaneNamespace,
+		}},
+	}
+	return clusterRole
+}
+
+func getConfigmapScriptPath(file string) string {
+	dir := os.Getenv("SCRIPT_PATH")
+	if dir != "" {
+		return path.Join(dir, file+".sh")
+	}
+	return path.Join("../../../../scripts", file+".sh")
+}
+
+func getConfigmapData() (string, error) {
+	fileName := "cleanup"
```

Comment:
Can we club both getConfigmapData and getConfigmapScriptPath to single method?

Reply:
It's easier for Unit test cases. that is why we have isolated the functions.

Reply:
I don't think this is right approach I think we can write UTs for a method of many lines

Reply:
We can do it. but is it necessary ?

Reply:
Combined the function